### ISSUE TITLE
[Feature] Protect ressources to prevent accidential deletion/data loss

### ISSUE THREAD

**Issue by ChrisHubinger** [mariadb-operator/mariadb-operator#643]
**Is your feature request related to a problem? Please describe.**
We ran into the issue that due to an error during upgrading the operator (API version Change) we accidentally deleted our databases. It was clearly a mistake on our side, but loosing the kubernetes resource for a Database should imho not cause a Database to be simply dropped - or at least it this should be possible to configure that. 
This is a actually a show stopper for us in critical installations at the moment.

I searched if there is any option to protect database ressources so that the operator will never delete them regardless of Kubernetes ressources are deleted, but could not find any.

**Describe the solution you'd like**

* Add a "protected: true" property to critical Resources (Database etc) to make sure they will not be deleted by the finalizer. Similar to the resource protection e.g. terraform implements with: 
```
lifecycle {
    prevent_destroy = true
}
```

**Describe alternatives you've considered**
Not using operator for production environments

**Environment details**:
environment independent

My golang skills are limited but if you, consider such an option I'd be happy to come up with a PR for this (at least as a PoC implementation)

**Comment by tgruenert** [comment]
This would also important for us as we use ArgoCD Gitops. Deleting an App should not destroy data. As Kubernetes keeps the PVC for StatefulSets also Databases should be kept (configurable) during install / remove cycles.

**Comment by mmontes11** [comment]
Hey there! Thanks for bringing this up.

In the context of SQL-related resources i.e. `User`, `Grant` and `Database` we could introduce a `finalizerPolicy` field to define the finalizer logic: 
- `Delete`: Cleanup the SQL resource. This is the current behaviour and will stay as default.
- `Skip`: Skip the finalizing logic, and therefore avoid deleting the SQL resource.

```yaml
apiVersion: k8s.mariadb.com/v1alpha1
kind: Database
metadata:
  name: database
spec:
  finalizerPolicy: Delete
``` 

I recently had a conversation with someone on reddit that seemed to have a complex use case where skipping the termination logic would be valuable:
https://www.reddit.com/r/kubernetes/comments/1cyuefx/comment/l5cde7z/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Contributions are welcome!

**Comment by mmontes11** [comment]
To elaborate a bit more, this `finalizerPolicy`/`cleanupPolicy` can also be supported in `MariaDB` to decide whether the PVCs are cleaned up or not on `MariaDB` deletion:
- https://github.com/mariadb-operator/mariadb-operator/issues/633#issuecomment-2207407956

**Comment by ChrisHubinger** [comment]
I like the idea with the policy attribute - I'll do my best to find the time in the next weeks and to come up with a first PR for that.

**Comment by mmontes11** [comment]
This has been supported via https://github.com/mariadb-operator/mariadb-operator/pull/800

You can now prevent the operator from deleting SQL resources by setting `cleanupPolicy = Skip` to `User`, `Grant` and  `Database` resources. Thanks @ChrisHubinger for the contribution, very much appreciated! 🙏🏻 

This will be released in `v0.0.30`, stay tuned!
